### PR TITLE
Test results filter on result value

### PR DESCRIPTION
**Pull request by jeremyzitomer-usds**
## Related Issue or Background Info

- https://app.zenhub.com/workspaces/simplereport-better-data-team-605b43d51ea9f700119a48ee/issues/cdcgov/prime-simplereport/1027

## Changes Proposed

- added optional `result` parameter to getTestResults(Count) gql queries
- changed `TestEventRepository` to accept `Specification`s, using those for result searching now
- added generation of static metamodel classes (in `build.gradle`) for type-safe query building for `Specification`s
- added tests

## Checklist for Author and Reviewer
- [ ] Run this branch locally
- [ ] Confirm you can run the testResults() and testResultsCount() queries in the graphQL playground locally using different combinations of facilityId, patientId, and result value

### UI
- [ ] Any changes to the UI/UX are approved by design 
- [ ] Any new or updated content (e.g. error messages) are approved by design 

### Testing
- [ ] Includes a summary of what a code reviewer should verify

### Changes are Backwards Compatible
- [ ] Database changes are submitted as a separate PR
  - [ ] Any new tables that do not contain PII are accompanied by a GRANT SELECT to the no-PHI user
  - [ ] Any changes to tables that have custom no-PHI views are accompanied by changes to those views
- [ ] GraphQL schema changes are backward compatible with older version of the front-end

### Security
- [ ] Changes with security implications have been approved by a security engineer (changes to  authentication, encryption, handling of PII, etc.)
- [ ] Any dependencies introduced have been vetted and discussed

## Cloud
- [ ] DevOps team has been notified if PR requires ops support
- [ ] If there are changes that cannot be tested locally, this has been deployed to our Azure `test` environment for verification

### CODE REVIEW

**Comment by jeremyzitomer-usds** [comment]
I guess @nullflux @nickclyde please ignore the Sonar failure? It is lying to you about 0% test coverage!! How mean.

File: backend/src/main/java/gov/cdc/usds/simplereport/service/TestOrderService.java [hunk 1]
```diff
@@ -81,27 +94,53 @@ public TestOrderService(
...
-  public List<TestEvent> getTestEventsResults(UUID facilityId, int pageOffset, int pageSize) {
-    Facility fac = _os.getFacilityInCurrentOrg(facilityId); // org access is checked here
-    return _terepo.getTestEventResults(fac.getInternalId(), PageRequest.of(pageOffset, pageSize));
-  }
-
-  public int getTestResultsCount(UUID facilityId) {
-    Facility fac = _os.getFacilityInCurrentOrg(facilityId); // org access is checked here
-    return _terepo.getTestResultsCount(fac.getInternalId());
+  // Specifications filters for queries
+  private Specification<TestEvent> buildTestEventSearchFilter(
```
Comment (nickclyde) [inline backend/src/main/java/gov/cdc/usds/simplereport/service/TestOrderService.java:98]: Wow, this is some fancy stuff, I've never seen this `Specification` class before. Not totally sure I understand how it works, but I like it better than having a bunch of different methods in the Service, great work!

File: backend/src/main/java/gov/cdc/usds/simplereport/db/model/TestOrder.java [hunk 2]
```diff
@@ -32,11 +31,9 @@
...
   private OrderStatus orderStatus;
 
-  // strictly speaking, this is no longer OneToOne since corrections could have
-  // more than one,
-  // but this is kept up-to-date with the latest one.
-  @Column(columnDefinition = "uuid")
-  private UUID testEventId; // id used directly without needing to load
+  @ManyToOne(fetch = FetchType.LAZY)
+  @JoinColumn(name = "test_event_id")
+  private TestEvent testEvent;
```
Comment (nickclyde) [inline backend/src/main/java/gov/cdc/usds/simplereport/db/model/TestOrder.java:36]: Nice improvement :+1:

File: backend/src/test/java/gov/cdc/usds/simplereport/service/TestOrderServiceTest.java [hunk 3]
```diff
@@ -874,32 +891,104 @@ void getTestEventsResults_pagination() {
     assertEquals(0, results_page3.size());
   }
 
+  @Test
+  @WithSimpleReportOrgAdminUser
+  void getTestEventsResults_filtering() {
```
Comment (nickclyde) [inline backend/src/test/java/gov/cdc/usds/simplereport/service/TestOrderServiceTest.java:896]: Thanks for the thorough tests!!

**Review by nickclyde — APPROVED** [review]
Just ran locally, everything looks to be working! Awesome work on this. Only note I have, is that there maybe shouldn't be any margin between the select box and the "Results" filter, so that it's more in line with the patient name search box:
image

File: backend/build.gradle [hunk 4]
```diff
@@ -61,6 +61,10 @@ dependencies {
     // SendGrid for Email
     implementation 'com.sendgrid:sendgrid-java:4.7.2'
 
+    // JPA Hibernate static metamodel generation
```
Comment (nullflux) [inline backend/build.gradle:64]: thanks for this 🙏

**Review by nullflux — APPROVED** [review]
Works on my machine and also LGTM! Next up: that one front end test...

**Comment by jeremyzitomer-usds** [comment]
> Just ran locally, everything looks to be working! Awesome work on this. Only note I have, is that there maybe shouldn't be any margin between the select box and the "Results" filter, so that it's more in line with the patient name search box:
> image

this should be fixed now @nickclyde !

**Comment by jeremyzitomer-usds** [comment]
https://user-images.githubusercontent.com/76444566/120844046-dbb95d00-c53c-11eb-8df1-03115de9766b.mov

@hmyers-cms @katiealoisi how's this look?

**Comment by jeremyzitomer-usds** [comment]
also the screen recording above has a small bug where the test results count doesn't update when selecting a result filter, but i've now updated it so that no matter what filters are applied, the results count reflects whatever is displayed on screen, accounting for filters.

**Comment by hmyers-cms** [comment]
The person search is a bit wonky full stop but that's definitely a separate design effort to improve. Right now, the full-width I think makes it a bit hard to tell what you're supposed to do next, because the "filter" button is so far away from the search box itself. Can we limit the max-width of that dropdown?

**Comment by jeremyzitomer-usds** [comment]
@hmyers-cms how about now?

https://user-images.githubusercontent.com/76444566/120859988-fdbdda00-c552-11eb-8c0e-64e06ea66e71.mov

**Review by hmyers-cms — APPROVED** [review]

**Pull request merged**